### PR TITLE
Resolve operator outside loop over value list in expression interpreter

### PR DESCRIPTION
**Pull request by lzeiming**
Reduce the time to reslove operator for issue #7015

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java [hunk 1]
```diff
@@ -167,6 +168,8 @@
     private final IdentityHashMap<LikePredicate, JoniRegexp> likePatternCache = new IdentityHashMap<>();
     private final IdentityHashMap<InListExpression, Set<?>> inListCache = new IdentityHashMap<>();
 
+    private final HashBasedTable<OperatorType, List<? extends Type>, ResolvedFunction> operatorCache = HashBasedTable.create();
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:171]: It can grow unbounded, so this doesn't seem a safe change.

Comment (lzeiming) [reply core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:171]: ExpressionInterpreter is created per expression, its lifecycle is short.
The existed inListCache and likePatternCache are also unbounded, so I follow the style.

I also considered adding a bounded thread-safe cache in MetadataManager, but I am not sure if it is proper.

**Comment by findepi** [comment]
Can you paste BenchmarkExpressionInterpreter from before the change and after?

**Comment by lzeiming** [comment]
### Before the change
```
Benchmark                                (inValuesCount)  Mode  Cnt     Score     Error  Units
BenchmarkExpressionInterpreter.optimize              200  avgt   20   108.974 ±  68.044  ms/op
BenchmarkExpressionInterpreter.optimize              500  avgt   20   157.350 ±  45.820  ms/op
BenchmarkExpressionInterpreter.optimize             1000  avgt   20   224.354 ±  10.850  ms/op
BenchmarkExpressionInterpreter.optimize             5000  avgt   20  1251.051 ± 113.997  ms/op
```

### After the change
```
Benchmark                                (inValuesCount)  Mode  Cnt    Score    Error  Units
BenchmarkExpressionInterpreter.optimize              200  avgt   20   35.808 ±  2.945  ms/op
BenchmarkExpressionInterpreter.optimize              500  avgt   20   96.284 ±  6.462  ms/op
BenchmarkExpressionInterpreter.optimize             1000  avgt   20  186.406 ± 17.595  ms/op
BenchmarkExpressionInterpreter.optimize             5000  avgt   20  967.069 ± 81.065  ms/op

```

**Comment by findepi** [comment]
The gain on large inValuesCount is smaller, as if the benefit of the change were diminishing (alleviated by JIT?).

What if we changed this line
https://github.com/trinodb/trino/blob/8fb4a95ab491d91dc82f9a566eaa519ddbc2097a/core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java#L621

to 
```
ResolvedFunction equalsOperator =  metadata.resolveOperator(OperatorType.EQUAL, types(node.getValue(), node.getValue()));
Boolean result = (Boolean) functionInvoker.invoke(equalsOperator, session, ImmutableList.of(value, inValue));
```

(note that I am replacing `types(node.getValue(), expression)` with `types(node.getValue(), node.getValue())` but this should be the same type)

Then you can pull `resolveOperator` before the `for` loop there.

Could you consider such a change and running your benchmark on it?

This can be further combine it with operator cache or not -- if the cache still provides visible gains.

**Comment by lzeiming** [comment]
### Resolve operator outside loop over value list. Without operator cache change.

```
Benchmark                                (inValuesCount)  Mode  Cnt    Score    Error  Units
BenchmarkExpressionInterpreter.optimize              200  avgt   20   33.364 ±  2.237  ms/op
BenchmarkExpressionInterpreter.optimize              500  avgt   20   91.773 ± 11.335  ms/op
BenchmarkExpressionInterpreter.optimize             1000  avgt   20  154.433 ±  4.217  ms/op
BenchmarkExpressionInterpreter.optimize             5000  avgt   20  812.423 ± 23.894  ms/op
```

### Combining with cache does not improve more.

```
Benchmark                                (inValuesCount)  Mode  Cnt    Score     Error  Units
BenchmarkExpressionInterpreter.optimize              200  avgt   20   32.025 ±   1.025  ms/op
BenchmarkExpressionInterpreter.optimize              500  avgt   20   91.298 ±   9.183  ms/op
BenchmarkExpressionInterpreter.optimize             1000  avgt   20  174.265 ±  11.066  ms/op
BenchmarkExpressionInterpreter.optimize             5000  avgt   20  907.463 ± 108.778  ms/op
```

**Review by findepi — APPROVED** [review]
Thanks!

CI failed https://github.com/trinodb/trino/issues/7085 -- sorry for that, no change required here.

@martint do you want to take a look?

**Review by martint — APPROVED** [review]

**Pull request merged**